Enable stronger cryptnono protections for public hubs + allow sharing shared directories between hubs (#3657)
2i2c-org/infrastructure: docs/howto/features/cryptnono.md

(howto:features:cryptnono)=
# Enable stronger anti-crypto abuse features for a hub

For JupyterHubs and BinderHubs that are broadly open to the public, cryptomining
attacks are the most common security threat. They take up resources and rack up
cloud costs. While most hubs can get away with gating access to them via some
kind of login restriction, for a subset of hubs this is not ideal (for equity and
access reasons).

For those cases, we enable a stronger deployment of [cryptnono](https://github.com/cryptnono/cryptnono).
The cryptnono project (in use on mybinder.org as well) helps detect and kill cryptominers via
various 'Detectors'.

## Detectors

Cryptnono currently has two primary detectors:

1. A detector for the [monero](https://www.getmonero.org/) cryptocurrency, that is based on
   official guidance [from the monero project](https://blog.px.dev/detect-monero-miners/) on
   how to detect it. This is fairly safe and has a very low false positive rate, and requires
   no configuration. So by default, **this detector is enabled on all hubs**.

> 2. A detector (`execWhacker`) based on heuristics, where the full commandline used to execute a process (

Reviewer: This is capitalized as `execWhacker` but the chart accept `execwhacker` rather than `execWhacker`.

I think it would be good to consistently refer to one capitalization found in the chart config, or something clearly distinguished as written for human eyes, like `the exec whacker`.

Author: Yes, I'll do that. I think when writing cryptnono I initially had called it `execWhacker` before moving to all lowercase - I'll think about what's the most appropriate casing, and unify that (including possibly changing it in cryptnono itself).

Author: I've just moved everything to say `execwhacker` for now, I'll update it here if I change it in cryptnono

   regardless of how it is started) is used to detect if a process is likely crypto mining,
   and if so, immediately kill it. This relies on a tweakable config of banned strings to
   look for in the commandline of a process, and is [constantly being tweaked](https://github.com/jupyterhub/mybinder.org-deploy/security/advisories/GHSA-j42g-x8qw-jjfh).
   Since making the list of banned strings public would make ban evasion easy, the list of
   strings (and the method used to generate them) is encrypted. You can read more details
   about the specific method used by looking in the encrypted code file (`deployer/commands/generate/cryptnono_config/encrypted_secret_blocklist.py`)

>    in this repository.

Reviewer 2: ```suggestion
   in the `infrastructure` repository.
```

I don't think it's clear when you're on https://infrastructure.2i2c.org which repository 'this' one is.

   Since this detector may have a non-0 false positive rate, it is currently *not* enabled by
   default. However, eventually, once the config matures enough (and we have tested it enough),
   this would also be enabled by default everywhere. In the meantime, we only enable it for

>    *clusters* with any hub that allows unfettered external access.

Reviewer: I opened https://github.com/cryptnono/cryptnono/issues/31 about this. I think a criteria for 2i2c to enable execwhacker by default should be that there are metrics or logs to help us rule out or conclude cryptnono's execwhacker killed a process.

Author: @consideRatio indeed! I've now enabled container information reporting as well, so we will know what pod / namespace the kill took place in. We're also now collecting the metrics that cryptnono exposes, even though there is no grafana dashboard for that yet.

Author: I've updated the docs to match this, including a sample log entry.

```

It should immediately die, with a message saying `Killed`. This is a randomly generated test string, set up
in an unencrypted fashion in `helm-charts/support/values.yaml` under `cryptnono.detectors.execwhacker.configs`,
to enable testing by engineers and others.

## Looking at logs to understand why a process was killed by `execwhacker`

Cryptnono is deployed as a [daemonset](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/),
so there should be one pod per node deployed. It will log each process it kills, why it kills them, and if it
intentionally spares a process, why as well. So if a process is being killed due to this, you can look at logs
to understand why - it may also lead to more tweaking of the generator config.

1. Find the *node* in which the user server is running.

   ```bash
   kubectl -n <hub-name> get pod -o wide
   ```

   The `-o wide` will add an additional column, `NODE`, showing which node the pods are running in. Find the
   node of the user pod you care about.

2. Find the appropriate `cryptnono` pod for this node.

   ```bash
   kubectl -n support get pod \

>     --field-selector spec.nodeName=<name-of-node>\

Reviewer: ```suggestion
    --field-selector spec.nodeName=<name-of-node> \
```

   ```

   This tells us that the name of the process, as well as why it was killed.

   ```{note}
   We will eventually be able to see the [pod & namespace](https://github.com/cryptnono/cryptnono/issues/30) information
   of killed processes as well.
   ```

   ```{tip}
   To make the JSON easier to read, you can pipe the logs command to `jq`.
   ```

## Regenerating list of banned strings

Periodically, we will have to regenerate the list of banned strings to tune `cryptnono`,
by running the following command:

```bash
deployer generate cryptnono-secret-config
```

This will update the file `helm-charts/support/enc-cryptnono.secret.values.yaml` and re-encrypt it.

## Working on the banned strings generator

The banned strings generator is a fairly simple python script, present in `deployer/commands/generate/cryptnono_config/enc-blocklist-generator.secret.py`.
It's unencrypted and loaded by code in `deployer/commands/generate/cryptnono_config/__init__.py`. There is inline

> documentation in `encrypted_secret_blocklist.py`, but how does one hack on it?

Reviewer: ```suggestion
documentation in `enc-blocklist-generator.secret.py`, but how does one maintain it?
```

   ```

2. Work on the code as you wish - it's just a regular python file.

3. Re-encrypt it with `sops`

   ```bash
    sops --encrypt deployer/commands/generate/cryptnono_config/blocklist-generator.secret.py > deployer/commands/generate/cryptnono_config/enc-blocklist-generator.secret.py
   ```

4. Run `deployer generate cryptnono-secret-config` to test.

## Right to Replicate considerations

2i2c's [Right to Replicate](https://2i2c.org/right-to-replicate/) guarantees that communities can leave
with their hubs whenever they please, without any secret sauce. Cryptnono has two pieces of secret info
here:

1. The script to generate the secret config.
2. The secret config itself.

If a community wishes to leave and needs the config, we can make sure they know what the config is and
how to keep it updated - very similar to how we would handle passing on CILogon credentials or similar to
them. Since none of the code for cryptnono itself is secret, this does not conflict with right to replicate.

## Future work

`cryptnono` also exposes prometheus metrics about processes it has killed. We currently do not collect these,

> but we should enable collection so we can

Reviewer 3: I think it misses a few closing words :D

Author: Ah yep! I've actually enabled collection of these now, and updated this section to say that we just don't have a grafana board yet.
